gomock/matchers.go

```diff
@@ -102,7 +102,21 @@ type eqMatcher struct {
 }
 
 func (e eqMatcher) Matches(x interface{}) bool {
-	return reflect.DeepEqual(e.x, x)
+	// In case, some value is nil
+	if e.x == nil || x == nil {
+		return reflect.DeepEqual(e.x, x)
+	}
+
+	// Check if types assignable and convert them to common type
+	x1Val := reflect.ValueOf(e.x)
+	x2Val := reflect.ValueOf(x)
+
+	if x1Val.Type().AssignableTo(x2Val.Type()) {
```

Comment:
I think for aliased types (`type A []string`) it doesn't matter, but if we are talking about interfaces I think there may be an issue here.

```
type Foo interface {
  Meow() string
}

type Baz struct {}

func (b Baz) Meow() string {
  return "meow"
}

type Boomer interface {
  Boom(Foo)
}

func TestBoom(t *testing.T) {
  // omitted for brevity
  var bazzer Baz = newBazzer()
  mockBoomer.EXPECT().Boom(gomock.Eq(bazzer)).Times(1)
  var fooer Foo = Foo(bazzer)
  mockBoomer.Boom(fooer) // Foo not assignable to Baz
}
```

Reply:
Ah ok, I think I understand. In the case of an interface the object will be resolved to its underlying type which must be assignable to the expected type. Thanks for the demo!